### DISCUSSION TITLE
Copilot in neovim gets stuck in setup/auth

### DISCUSSION THREAD

**Discussion by tedmiddleton** [community/community#50263] [Copilot Conversations]
### Select Topic Area

Bug

### Body

I'm using neovim 0.8.2 in Ubuntu 22.04.1 LTS running in WSL in Windows 11. I've installed a version 18.15.0 of node.js (because copilot wants a much newer version of node.js than what comes with Ubuntu 22.04). 

I've installed copilot with vimplug and when I type `:Copilot status` I get

```
Copilot: Not authenticated. Invoke :Copilot setup
```

When I run `:Copilot setup` I get

```
First copy your one-time code: XXXX-XXXX
Press ENTER to open GitHub in your browser
```

I press ENTER, and...

```
First copy your one-time code: XXXX-XXXX
Press ENTER to open GitHub in your browser
Failed to open browser.  Visit https://github.com/login/device
Waiting (could take up to 5 seconds)
```

Ok, so it can't figure out how to open the page, but that's ok? I navigate to that page, input the XXXX-XXXX one-time code I've been given, and auth seems to work in the browser:

![image](https://user-images.githubusercontent.com/295132/225716884-398931cc-ccc0-4c0c-b716-8873738073a9.png)

But, neovim is stuck hanging at 

```
First copy your one-time code: XXXX-XXXX
Press ENTER to open GitHub in your browser
Failed to open browser.  Visit https://github.com/login/device
Waiting (could take up to 5 seconds)
```

The 5 seconds (and the cake) is clearly a lie here :D 

neovim is unresponsive and I have to break with CTRL-C:

```
First copy your one-time code: XXXX-XXXX
Press ENTER to open GitHub in your browser
Failed to open browser.  Visit https://github.com/login/device
Waiting (could take up to 5 seconds)
Interrupt: Press ENTER or type command to continue
```

and... `:Copilot status` still says

```
Copilot: Not authenticated. Invoke :Copilot setup
```

**Comment by tedmiddleton** [comment]
There seems to be a couple of similar discussions, but they all seem to involve upgrading Powershell (I'm using WSL not Powershell) or they just don't have resolutions.

**Comment by tedmiddleton** [comment]
Is there anything I can run on my machine to help isolate the problem?

**Comment by tedmiddleton** [comment]
```
      else
        call s:Echo("Could not find browser.  Visit " . uri)
      endif
      call s:Echo("Waiting (could take up to 5 seconds)")
      let request = copilot#Request('signInConfirm', {'userCode': data.userCode}).Wait()
    finally
      if exists('mouse')
        let &mouse = mouse
      endif
    endtry
```

So it's hanging in `copilot.vim` at line ~678 in `s:commands.setup()`. I guess that it's making an http request of the github server? I've entered the code at `http://github.com/login/device` but it just never returns from `copilot#Request('signInConfirm')`. 

Any clues?

**Comment by tedmiddleton** [comment]
One thing I noticed is that when I install `wslview`, the browser invocation seems to work. So you might add that as a note for neovim/wsl users.

**Comment by tedmiddleton** [comment]
So, just looking at how you've got this set up, it looks like you're starting an LSP server implemented in node in agent.js. When neovim gets stuck waiting, it's stuck in a function in agent.vim called `RequestWait()`, which presumably is waiting for something to arrive from the LSP server? I can see that the server subprocess is still running, but nothing is happening.

**Comment by tedmiddleton** [comment]
Just to be completely clear, I have every function in agent.vim instrumented, and the last one to get called is `RequestWait()`. Nothing happens after that. 

Any clues? It looks like agent.js is heavily minified/obfuscated. I don't really want to have to trudge around in javascript but it seems like that might be necessary.

**Comment by tedmiddleton** [comment]
Ok, well the adventure continues. I turned on lsp logging in neovim to TRACE level and here's what I see when I try to execute `:Copilot setup`:

```
[DEBUG][2023-03-21 03:36:05] .../lua/vim/lsp.lua:1381   "LSP[copilot]"  "client.request"        20      "checkStatus"   {  options = {    localChecksOnly = true  }}    <function 1>    3
[DEBUG][2023-03-21 03:36:05] .../vim/lsp/rpc.lua:285    "rpc.send"      {  id = 16,  jsonrpc = "2.0",  method = "checkStatus",  params = {    options = {      localChecksOnly = true    }  }}
[DEBUG][2023-03-21 03:36:05] .../vim/lsp/rpc.lua:388    "rpc.receive"   {  id = 16,  jsonrpc = "2.0",  result = {    status = "NotSignedIn"  }}
[DEBUG][2023-03-21 03:36:07] .../lua/vim/lsp.lua:1381   "LSP[copilot]"  "client.request"        20      "checkStatus"   vim.empty_dict()        <function 1>    3
[DEBUG][2023-03-21 03:36:07] .../vim/lsp/rpc.lua:285    "rpc.send"      {  id = 17,  jsonrpc = "2.0",  method = "checkStatus",  params = vim.empty_dict()}
[DEBUG][2023-03-21 03:36:07] .../vim/lsp/rpc.lua:388    "rpc.receive"   {  id = 17,  jsonrpc = "2.0",  result = {    status = "NotSignedIn"  }}
[DEBUG][2023-03-21 03:36:08] .../lua/vim/lsp.lua:1381   "LSP[copilot]"  "client.request"        20      "signInInitiate"        vim.empty_dict()        <function 1>    3
[DEBUG][2023-03-21 03:36:08] .../vim/lsp/rpc.lua:285    "rpc.send"      {  id = 18,  jsonrpc = "2.0",  method = "signInInitiate",  params = vim.empty_dict()}
[DEBUG][2023-03-21 03:36:08] .../vim/lsp/rpc.lua:388    "rpc.receive"   {  id = 18,  jsonrpc = "2.0",  result = {    expiresIn = 899,    interval = 5,    status = "PromptUserDeviceFlow",    userCode = "XXXX-XXXX",    verificationUri = "https://github.com/login/device"  }}
[DEBUG][2023-03-21 03:36:32] .../lua/vim/lsp.lua:1381   "LSP[copilot]"  "client.request"        20      "signInConfirm" {  userCode = "XXXX-XXXX"}      <function 1>    3
[DEBUG][2023-03-21 03:36:32] .../vim/lsp/rpc.lua:285    "rpc.send"      {  id = 19,  jsonrpc = "2.0",  method = "signInConfirm",  params = {    userCode = "XXXX-XXXX"  }}
```
 
The stuff from 03:36:05 to 03:36:08 happens immediately and then waits for me to `Press ENTER to open GitHub in your browser`. Then at 03:36:32 I pressed ENTER and those last 2 lines were emitted and my browser window opened. I pasted the device id into the browser web page and then clicked the "authorize" button. But there's nothing else that `copilot.js` sends over the LSP channel.

So the problem isn't in the vimscript - it's either in the neovim LSP machinery, `copilot.js`, or the copilot server.

The next step, I think, is to inspect and instrument `copilot.js`.

**Comment by tedmiddleton** [comment]
Something I didn't notice the last time I looked at neovim's lsp log:
```
[WARN][2023-03-21 05:00:11] ...lsp/handlers.lua:113	"The language server copilot triggers a registerCapability handler despite dynamicRegistration set to false. Report upstream, this warning is harmless"
```
It says it's "harmless" - is it, I wonder?

**Comment by tedmiddleton** [comment]
And, strangely, it seems to work now? I can only assume that something was fixed on the server side? At any rate, I'm closing this bug now that I can authenticate.

**Comment by RaulSeganfreddo** [comment]
@tedmiddleton Hey man, I'm having the exact same issue, do you have a solution? Thanks!

**Reply by tedmiddleton** [reply]
@RaulSeganfreddo  - I'm not sure what happened. It might have been that something was wrong with my ssh keys? When github replaced their ssh certificate on March 24, I had to reset my known hosts and it started working then.

**Reply by devvspaces** [reply]
I had to use neovim, it doesn't work with vim

**Reply by mikaeilorfanian** [reply]
I'm using neovim and wsl2, but get stuck after auth is done in the browser. Unfortunately none of the suggested solutions work. Copilot works when I use it within Vs code on windows. I thought this was a WSL networking issue, but all other networking ops work from within WSL.

**Comment by mikaeilorfanian** [comment]
For those who encounter this issue in future. I'm using Ubuntu with WSL2, neovim 0.9, and copilot.vim 1.8.4, node 18.16. I was stuck "waiting for 5 seconds" after auth was done on Github's page. I manually opened the URL copilot.vim outputs after doing the `:Copilot setup` command.
@tedmiddleton has done some detailed investigative work above on why copilot.vim gets stuck and he gave a very important clue: the copilot.vim waits or dependds on the LSP. So, after I opened a Python module which lazy loads my LSP, then the copilot.vim setup went through in less than 5 seconds after auth on Github's page.
Thanks a bunch @tedmiddleton for the detailed debugging.

**Reply by hf-xz** [reply]
This works for me, open a random python file and then run `Copilot auth`.

**Comment by rayfiyo** [comment]
@mikaeilorfanian is right, just follow the output and access the URL.
"Failed to open browser. Visit https://github.com/login/device"

Alternatively, you can install a browser such as Chrome in your WSL, which will automatically launch the browser.

**Comment by vimkim** [comment]
I tried every method above and it still won't work...

**Reply by mikaeilorfanian** [reply]
Yeah it used to work for me, but then it stopped. I ended up using neovim on Ubuntu and PyCharm on windows. Sounds like not many people who use vim + wsl run into this issue, otherwise somebody would fix it. It definitely seems to be a client issue. On VSCode there's never any issues logging in. I ran into an issue logging in on PyCharm once after testing it about 10 times. Using WSL + neovim it always breaks. Neovim on Ubuntu works. Good luck, if you figure anything out please share. Thanks.

**Reply by F18** [reply]
I have also tried every method above and none are working for me. I am using wsl2 (ubuntu 22.04) + neovim.

**Comment by insanityinside** [comment]
Something useful to know - this also hangs if you're using an MITM proxy such as Zscaler with a custom CA certificate, in the same way that using VSCode with WSL has the same issue with connectivity to the CoPilot service.

I was able to successfully complete setup using similar information to the VS code troubleshooting. I solved the login hang by setting the following environment variable (customise to your individual setup/shell/cert filename, examples given for Debian-based and RHEL-based, add in an appropriate startup script etc):

### Ubuntu/Debian
export NODE_EXTRA_CA_CERTS=/usr/local/share/ca-certificates/ZscalerRootCertificate-2048-SHA256.crt
 
### RHEL/CentOS/Rocky/RedHat
export NODE_EXTRA_CA_CERTS=/etc/pki/ca-trust/source/anchors/ZscalerRootCertificate-2048-SHA256.crt

Hope this helps someone also trying to solve the same issue.

**Reply by edis-netlight** [reply]
You sir are a hero!

**Reply by zekiz774** [reply]
Thank you so much. This fixed an issue with WSL/Zscaler that I have for months already

**Comment by twoscomp** [comment]
Running Ubuntu on WSL2 under Windows 11 + neovim. I also previously had this issue where the auth dialog would never dismiss. I recently switched my default browser from Chrome/Edge to Firefox and the problem resolved. Could the issue be with the chromium?

**Reply by gblais** [reply]
This solved the issue for me too. Thanks.

**Comment by filip-figiel-alma** [comment]
I had the same problem with Ubuntu/WSL2. After stopping my linux vm via `wsl.exe --shutdown` and trying again, the authentication worked

**Reply by ysw421** [reply]
its worked! :) 👍

**Reply by viscousbiscuit** [reply]
worked for me too, I also installed firefox first as stated above `sudo apt install firefox`.

**Reply by roskee** [reply]
It worked for me too. I first installed firefox with `sudo snap install firefox` then did `wsl.exe --shutdown` on cmd. then launched a wsl terminal and it worked on the first try.

**Reply by Dixiat** [reply]
worked for me too! I'm using edge by the way.

**Reply by gelegele** [reply]
worked for me too.  Thanks

**Reply by rschlack** [reply]
This method still works when the dialog is stuck.

**Comment by ghry5** [comment]
Only way I was able to get it to work:
- Install firefox inside wsl ubuntu: `sudo apt install firefox`
- open firefox inside wsl ubuntu -> (firefox opens inside wslg)
- open neovim & start `Copilot setup/auth`
- manually open https://github.com/login/device inside the open firefox window
- login to github and enter/paste code.

**Comment by eidsnessje-nih** [comment]
Mine doesn't work either, using vim in a "Git for Windows" console. But why would it? It's doing too much. Logging into things is way too complicated anymore, and adding it in as a plugin to something is just kind of insane. In general though, it's not like getting logged into things with 2 factor and tokens etc. is so impossible.   Eventually this thing is going to get us a JWT that it will use - If any of you could just find me where it's saving (and subsequently looking for) that, I can get my own.  I can tell it's using node, as well, so if we could just get the node login commands to run outside of vim, instead of trying to make the plugin understand what's going on, that's another way we wouldn't have to expect it to work inside of vim.

**Reply by eidsnessje-nih** [reply]
I have gotten past the place I was stuck, in the attempt to get a device code. Before, it would say that this could take over 10 seconds,  but in 10 seconds or less, it would stop waiting, and return to vim's previous mode.  At that point, entering in the code, waiting for additional instructions and permitting the application in the browser would not do anything, anymore.

For reasons I do not fully understand, my situation improved after I added my firefox.exe location as g:copilot_browser, in $MYVIMRC, and restarted vim & the ":Copilot setup" process.

    let g:copilot_browser = "C:\\Program Files\\Mozilla Firefox\\firefox.exe"

That was based on what I saw under the "Copilot#Browser" function. I was hoping it would learn that it could start a browser after all. Well, that did not start firefox, but it changed something else, indirectly. Somehow after giving that hint about a browser, it could now wait more than the 10 seconds for me to complete the device code authorization. I followed the instructions in the browser, and waited, then it fell right into shape, and the system is working out.

**Comment by surgiie** [comment]
Ive tried every method here. Nothing seems to work for me :/

**Reply by meshy** [reply]
Perhaps try this: https://github.com/orgs/community/discussions/152171#discussioncomment-12324088

**Comment by ilan-schemoul** [comment]
Not a single thing worked. What a shame to be a company who makes money out of people hosting open source software on your platform yet refusing to open source anything and when it's broken  not fixing it or anything. Everyone complain and need weird workaround. Either open source and I'll fix it or close source but make it work. Can't fail on both fronts.

**Reply by ilan-schemoul** [reply]
Oh ok after reinstalling firefox and restarting wsl multiple times for some reason and after like 15 attempts it finally randomly worked.
Best way to lose customer trust before we even tried anything.
Meanwhile cursor is known to have a flawless experience

**Comment by kurojs** [comment]
If nothing else works and you're using WSL, what worked for me was installing Neovim on Windows, authenticating with :Copilot auth, then copying the apps.json file from C:\Users\<YourUsername>\AppData\Local\github-copilot\apps.json to ~/.config/github-copilot/ in WSL. After that, I ran :Copilot auth info in Neovim on WSL to verify it was working.  🙌

**Reply by ericwarren** [reply]
This worked for me.  Thank you!

**Reply by ted-mundy** [reply]
just tried this, and it didn't work for me.

```
 pwd && cat apps.json
> /home/ted/.config/github-copilot
> {"github.com:Iv1.DONTKNOW":{"oauth_token":"WHATINEEDTO","user":"ted-mundy","githubAppId":"Iv1.CENSORHERE"}}

:Copilot status
> Copilot: Not authenticated. Invoke :Copilot setup

:Copilot auth info
> First copy your one-time code ...
```

after moving the file to the WSL machine, it still seems that nvim "remembers" the login, as there is no `apps.json` on windows, but it is still authenticated?

edit: i've given up with this - switched over to https://github.com/zbirenbaum/copilot.lua

**Comment by mariusfa** [comment]
Why is this closed. Still a same problem with most recent copilot and neovim(0.11)?
Hangs on add device...

**Comment by surgiie** [comment]
I have opened an actual issue on the repo: https://github.com/github/copilot.vim/issues/170

Kinda crazy this is not fixed yet. :D
